kindlyops/mappamundi
# Session save

Fixes:

https://kindlyops.atlassian.net/browse/HAV-72

https://kindlyops.atlassian.net/browse/HAV-115

# Description

This enables the survey page to be refreshed mid-survey for both Likert and Ipsative surveys. What made this problem hard was the fact that the survey uses Elm's Tagged Unions:

```
type Survey
    = Ipsative IpsativeSurvey
    | Likert LikertSurvey
```

... and also uses Zippers for the questions:
```
type alias IpsativeSurvey =
    { metaData : SurveyMetaData
    , pointsPerQuestion : Int
    , numGroups : Int
    , questions : Zipper.Zipper IpsativeQuestion
    }
```

Which Elm doesn't allow through its ports or during initialization.

I haven't tested this with the original problem (being kicked out from keycloak). But I expect that it should work.

Also, there's definitely room to re-factor, split, and simplify the code.

## How to test

1. `git checkout session-save`
2. `docker-compose up`
3. Login and go to the survey page
4. Pick a survey, fill in some answers
5. Refresh your browser

The page should pick up where you left off.
